src/usbhs/mod.rs

```diff
@@ -0,0 +1,434 @@
+
+        // Following WCH C code.... unclear why clr_all is not also cleared here
+        r.ctrl().modify(|w| w.set_reset_sie(false));
+        h.ctrl().write(|w| w.set_phy_suspendm(true));
+
+        r.ctrl().write(|w| {
+            w.set_int_busy(true);
+            w.set_dma_en(true);
+            // TODO: HS, Configurable?
+            w.set_speed_type(SpeedType::HIGHSPEED);
+        });
+
+        r.int_en().write(|w| {
+            w.set_dev_nak(false);
+            w.set_fifo_ov(true);
+            w.set_setup_act(true);
+            w.set_suspend(true);
+            w.set_bus_rst(true);
+            w.set_transfer(true);
+        });
+
+        Self {
+            phantom: PhantomData,
+            allocator,
+            // 0 is reserved for the control endpoint
+            next_ep_addr: 1,
+        }
+    }
+
+    fn alloc_ep_address(&mut self) -> u8 {
+        if self.next_ep_addr as usize >= MAX_NR_EP {
+            panic!("ep addr overflow")
+        }
+
+        let addr = self.next_ep_addr;
+        self.next_ep_addr += 1;
+        addr
+    }
+
+    fn alloc_endpoint<D: Dir>(
+        &mut self,
+        ep_type: EndpointType,
+        max_packet_size: u16,
+        interval_ms: u8,
+        dir: Direction,
+    ) -> Result<Endpoint<'d, T, D>, EndpointAllocError> {
+        let ep_addr = self.alloc_ep_address();
+        let data = self.allocator.alloc_endpoint(max_packet_size)?;
+
+        Ok(Endpoint::new(
+            EndpointInfo {
+                // todo fix in embassy usb driver ep_addr should be u8 with top bit unset
+                addr: EndpointAddress::from_parts(ep_addr as usize, dir),
+                ep_type,
+                max_packet_size,
+                interval_ms,
+            },
+            data,
+        ))
+    }
+}
+
+impl<'d, T: Instance, const NR_EP: usize> embassy_usb_driver::Driver<'d> for Driver<'d, T, NR_EP> {
+    type EndpointOut = Endpoint<'d, T, Out>;
+    type EndpointIn = Endpoint<'d, T, In>;
+    type ControlPipe = ControlPipe<'d, T>;
+    type Bus = Bus<'d, T>;
+
+    fn alloc_endpoint_out(
+        &mut self,
+        ep_type: EndpointType,
+        max_packet_size: u16,
+        interval_ms: u8,
+    ) -> Result<Self::EndpointOut, EndpointAllocError> {
+        self.alloc_endpoint(ep_type, max_packet_size, interval_ms, Direction::Out)
+    }
+
+    fn alloc_endpoint_in(
+        &mut self,
+        ep_type: EndpointType,
+        max_packet_size: u16,
+        interval_ms: u8,
+    ) -> Result<Self::EndpointIn, EndpointAllocError> {
+        self.alloc_endpoint(ep_type, max_packet_size, interval_ms, Direction::In)
+    }
+
+    fn start(mut self, control_max_packet_size: u16) -> (Self::Bus, Self::ControlPipe) {
+        let d = T::dregs();
+
+        let ep0_buf = self.allocator.alloc_endpoint(control_max_packet_size).unwrap();
+        d.ep0_dma().write_value(ep0_buf.addr() as u32);
+        d.ep_max_len(0).write(|w| w.set_len(64));
```

Comment:
this should probably be `control_max_packet_size`. not that it’ll be anything other than 64 unless ls/fs support is added later